### ISSUE TITLE
Realm.init(configuration:) throws Obj-C exceptions that can't be caught in Swift

### ISSUE THREAD

**Issue by jakob** [realm/realm-cocoa#3002]
I want to catch errors when opening a Realm file using do/catch:

```swift
let realm : Realm
do {
    realm = try Realm(configuration: config)
} catch let error {
    print("Realm could not be opened:", error)
}
```

However, when an error occurs, for example because there is no migration block, but a migration is necessary, Realm throws an Objective C Exception that I can't catch in Swift, leading the app to crash instead of handling the error gracefully.

**Comment by mrackwitz** [comment]
We have different concepts for differentiating whether we throw an errors or raise an exception or more severe measures. It might be hard sometimes to draw that line or to follow our reasoning about that. But in general the idea behind is the following:

* The method is throwing / takes an NSError pointer, when it can fail because of erroneous inputs, which can be handled gracefully and where it's in the frameworks responsibility to process the inputs and it there was no possibility for the developer to determine that these inputs could be wrong.
* An exception is thrown when the developer did something wrong. The program is expected to be in a state, where it can't be continued like that. The problem should be caught in development and should be hard to miss. An exception is also thrown if a shortcut method was used, which doesn't allow to propagate a graceful error or no error pointer was provided from Objective-C. They can be stilled caught from Objective-C as a last resort for special circumstances.
* An assert occurs when something in the underlying storage engine is broken and/or no appropriate error / exception handling is in place on the upper layers. An assert leads inevitable to a crash.

In your case, I guess you can use [`schemaVersionAtPath(_:encryptionKey:error:)`](https://realm.io/docs/swift/latest/api/Functions.html#/s:F10RealmSwift19schemaVersionAtPathFTSS13encryptionKeyGSqCSo6NSData_5errorGVSs33AutoreleasingUnsafeMutablePointerGSqCSo7NSError___GSqVSs6UInt64_) to validate the current schema version. On base of that you should be able to judge how you want to proceed. If that's not the case, it would be helpful if you could explain your use-case a little bit better.

**Comment by jakob** [comment]
Here's my use case: *I want to show reasonable error messages to the user.*

Checking if the schema version matches is not sufficient because a matching schema version doesn't guarantee that the call will succeed.

My app is currently under development, and the schema is constantly changing. I do not set up migrations because the app is not yet public, and I do not increment the schema version yet. So it often occurs that the Realm database does not match the model. In that case I want to show a warning to the user, instead of having the app crash.

In my opinion this does not fall into the category of "developer error". There are any number of reasons why a Realm file you are trying to open may not match the structure you are expecting (eg. maybe someone created the realm file with a different app?). I want to be able to catch this error somehow.

The way I see it, opening a Realm file with unexpected schema falls squarely into your first category "erroneous inputs".

(I'm not asking for help; I know how to catch the error in Obj-C; I'm suggesting to change the Realm-Cocoa API in this regard.)

**Comment by mrackwitz** [comment]
At some places, we might have preferred the second over the first, because let's face it: NSError pointers are ugly from an API-design point of view. Also the propagation through the framework's callstack and chaining of failable methods is much easier to implement for exceptions than for NSError pointers. The line between them is not necessarily fixed and we're open to re-evaluate certain decisions, especially where we built our Swift binding as a wrapper around the Objective-C binding from the very beginning of the new language, where there was no error handling concept in place yet. So there might be still very valid use-cases, where it would be reasonable to have at least a failable alternative method available in Objective-C, which might then be used exclusively in Swift. On the one hand doing that would move the responsibility to decide over putting error handling in place or asserting instead by using `try!` in the hands of the developer and raise awareness for what could go wrong. On the other hand the API gets a little more clumsy again and not as crisp as we generally prefer. It is still a trade-off.

That being said, here in that case, we already have a failable method. I'll bring up your concern with the rest of the team and discuss that. 

(We tend to use Help lately as a first stage label, until we have decided upon in which direction this is moving. Your feedback is very welcome.)

**Comment by jpsim** [comment]
> My app is currently under development, and the schema is constantly changing. I do not set up migrations because the app is not yet public, and I do not increment the schema version yet. So it often occurs that the Realm database does not match the model. In that case I want to show a warning to the user, instead of having the app crash.

I think the appropriate solution for this is to allow configuring a Realm file to be deleted when a migration is required, which we're tracking as #1692.

There's also the [`schemaVersionAtPath(...)`](https://realm.io/docs/swift/latest/api/Functions.html#/s:F10RealmSwift19schemaVersionAtPathFTSS13encryptionKeyGSqCSo6NSData_5errorGVSs33AutoreleasingUnsafeMutablePointerGSqCSo7NSError___GSqVSs6UInt64_) function that could be useful to you here.

**Comment by jpsim** [comment]
> Checking if the schema version matches is not sufficient because a matching schema version doesn't guarantee that the call will succeed.

Oops, sorry for suggesting that function, I skimmed this part of your comment.

---

Reading through this again, I think there might be some valid use cases for determining if a Realm at a given path requires a migration. This might be something we want to expose in a dynamic API.

**Comment by jakob** [comment]
@jpsim Yeah, I actually don't want to delete the Realm file when an error occurs. I just want to detect that an error happened and handle it gracefully. If there's a way I can check if the call would fail ahead of time that would be fine too.

Anyway, I came up with a workaround that lets me catch the exception and handle it gracefully. It probably leaks memory, since AFAIK neither Swift nor Obj-C ARC are exception-safe, but I prefer leaking memory and showing an error message to crashing.

**Comment by jpsim** [comment]
> Yeah, I actually don't want to delete the Realm file when an error occurs. I just want to detect that an error happened and handle it gracefully. If there's a way I can check if the call would fail ahead of time that would be fine too.

Understood, and sorry for the confusion. I'd recommend you catch the "migration needed" exception as you mentioned as a work around, even though this will likely lead some memory.

I *might* be convinced to make this exception into a recoverable error, because there are *some* cases in which it's not a programmer error, like if you're building an app that interacts with Realm files that you don't control (which is the case for the app you're building), but I really don't think that's a common situation to be in.
